lib/internal/modules/esm/default_resolve.js

```diff
@@ -51,17 +51,50 @@ if (experimentalWasmModules)
 if (experimentalJsonModules)
   extensionFormatMap['.json'] = legacyExtensionFormatMap['.json'] = 'json';
 
-function resolve(specifier, parentURL) {
+function defaultGetFormat({ url, isMain }, defaultGetFormat, loader) {
+  if (NativeModule.canBeRequiredByUsers(url)) {
+    return { format: 'builtin' };
+  }
+  const parsed = new URL(url);
+  if (parsed.protocol === 'data:') {
+    const [ , mime ] = /^([^/]+\/[^;,]+)(?:[^,]*?)(;base64)?,/.exec(parsed.pathname) || [ null, null, null ];
+    const format = ({
+      '__proto__': null,
+      'text/javascript': 'module',
+      'application/json': experimentalJsonModules ? 'json' : null,
+      'application/wasm': experimentalWasmModules ? 'wasm' : null
+    })[mime] || null;
+    return { format };
+  } else if (parsed.protocol === 'file:') {
+    const ext = extname(url.pathname);
+    let format;
+    if (ext === '.js' || ext === '') {
+      format = getPackageType(url.href) === TYPE_MODULE ? 'module' : 'commonjs';
+    } else {
+      format = extensionFormatMap[ext];
+    }
+    if (!format) {
+      if (experimentalSpeciferResolution === 'node') {
+        process.emitWarning(
+          'The Node.js specifier resolution in ESM is experimental.',
+          'ExperimentalWarning');
+        format = legacyExtensionFormatMap[ext];
+      } else {
+        throw new ERR_UNKNOWN_FILE_EXTENSION(ext, fileURLToPath(url));
+      }
+    }
+    return { format: format || null };
+  }
+}
+exports.defaultGetFormat = defaultGetFormat;
+
+function defaultResolve({ specifier, parentURL }, defaultResolve, loader) {
```

Comment:
I really liked the previous resolve signature of `resolve (specifier, { parentURL } (parentContext), defaultResolve)`, and it also aligns with the recent comments while ensuring that even if there are a large number of dependency specifiers they don't need their own objects (number of import statements >> number of modules).

Reply:
The previous signature was `resolve(specifier, parentURL, defaultResolve)`.

Reply:
perhaps with the talk about unified signatures we should put this in the non-file comments to get a unified place to compare all the signatures at once